glib/src/object.rs

```diff
@@ -288,6 +288,75 @@ pub trait Cast: ObjectType {
 
 impl<T: ObjectType> Cast for T {}
 
+// rustdoc-stripper-ignore-next
+/// Convenience trait mirroring `Cast`, implemented on `Option<Object>` types.
+///
+/// # Example
+/// ```ignore
+/// let widget: Option<Widget> = list_item.child();
+///
+/// // Without using `CastNone`
+/// let label = widget.unwrap().downcast::<gtk::Label>().unwrap();
+///
+/// // Using `CastNone` we can avoid the first `unwrap()` call
+/// let label = widget.downcast::<gtk::Label>().unwrap();
+/// ````
+pub trait CastNone: Sized {
+    type Inner;
+    fn downcast<T: ObjectType>(self) -> Result<T, Self>
```

Comment:
@pbor suggests to prefix all those functions in this trait with `then_`, that would make it extra clear that you are dealing with a cast from an optional type and not the ones from the Cast trait

Reply:
My suggestion was to suffix with `_opt()` but either seems fine to me as distinction, I somewhat prefer @pbor's

Reply:
I feel adding a prefix/suffix would defeat the purpose of the trait. It's a conviene trait to write less things. If we make the name as long as `.unwrap().downcast()`, the trait doesn't make sense anymore.

Reply:
My main concern here is that by using the same names you

  * make it likely that people use it accidentally and then handle type mismatches / non-existence of values wrong
  * if the value does not implement the trait, you get a compiler error list both traits and all their possible implementors, which is a lot

Reply:
> I feel adding a prefix/suffix would defeat the purpose of the trait. It's a conviene trait to write less things. If we make the name as long as `.unwrap().downcast()`, the trait doesn't make sense anymore.

Adding then_ as a prefix is waaay to less code than what we currently have and I completely agree with the concerns raised by Sebastian here.

Reply:
I've chosen the prefix `and_`. Why?

Because `then` is used when the user should provide a function. See for example `Option::and_then` or `bool::then`.

The `and_` prefix is used when you the user wants to map an option to another option, and the process returns itself an option.
I think the concept maps well: 
```rust
// Option::and
(None::<usize>).and(Some(2u8)) -> None::<u8>
// CastNone::and_downcast
(None::<gtk::Widget>).and_downcast::<gtk::Button>() -> None::<gtk::Button>
```

Reply:
I like the naming. @bilelmoussaoui?

Could maybe also be implemented on `Result` now instead of just on `Option`?

Reply:
Going from a Result to an Option only requires calling `result.ok()`, so I believe having a shortcut method on Result it's unnecessary. It would also require a new trait, because every method `and_downcast*` would need to return a `Result` instead of an `Option`, to keep consistency with the other `and_*` functions defined on `Result`.

Reply:
Looks good to me!